File: frontend/src/types/transactions.ts
export type TransactionType = "income" | "expense" | "transfer";

export interface Transaction {
    id: number;
    name: string;
    amount: number;
    category: string;
    transaction_type: TransactionType;
    description: string | null;
    transaction_date: string;
    created_at: string;
    last_updated: string | null;
    merchant: string | null;
}

export interface PaginatedTransactionsResponse {
    items: Transaction[];
    page: number;
    page_size: number;
    total_items: number;
    total_pages: number;
    has_next: boolean;
    has_previous: boolean;
}

export interface TransactionSort {
    sort_by: "name" | "amount" | "category" | "transaction_type" | "transaction_date";
}

export interface TransactionSortOrder {
    sort_order: "asc" | "desc";
}

export interface TransactionCreate {
    name: string;
    amount: number;
    category: string;
    transaction_type: TransactionType;
    description: string | null;
    transaction_date: string;
    merchant: string | null;
}

export interface TransactionUpdate {
    name?: string;
    amount?: number;
    category?: string;
    transaction_type?: TransactionType;
    description?: string | null;
    transaction_date?: string;
    merchant?: string | null;
}

File: backend/app/services/transactions.py
from app.models.models import Transaction
from app.schemas.schemas import TransactionSummary
from app.utils.pagination import get_total_pages, get_offset


def build_summary(transactions: list[Transaction]) -> TransactionSummary:
    summary = {
        "total_income": 0,
        "total_expense": 0,
        "total_transfer": 0,
        "net_balance": 0,
        "transaction_count": 0
    }
    
    for transaction in transactions:
        type_lower = transaction.transaction_type.lower()
        if type_lower == "income":
            summary["total_income"] += transaction.amount
        elif type_lower == "expense":
            summary["total_expense"] += transaction.amount
        elif type_lower == "transfer":
            summary["total_transfer"] += transaction.amount
            
    summary["net_balance"] = summary["total_income"] - summary["total_expense"]
    summary["transaction_count"] = len(transactions)
    
    return TransactionSummary(**summary)

def build_pagination(transactions: list[Transaction], page: int, page_size: int, total_items: int) -> dict:
    total_pages = get_total_pages(total_items=total_items, page_size=page_size)
    
    return {
        "items": transactions,
        "page": page,
        "page_size": page_size,
        "total_items": total_items,
        "total_pages": total_pages,
        "has_next": page < total_pages,
        "has_previous": total_pages > 0 and page > 1
    }